gap/congruences/conglatt.gi

```diff
@@ -240,136 +229,260 @@ SEMIGROUPS.LatticeOfXCongruences := function(S, type_string, record)
-                                         "CongruencesOfSemigroup"));
-    set_func(S, congs);
-  fi;
 
-  SEMIGROUPS.OptionsRec(S).report := report;
-  # Objectify the result
-  lattice := Objectify(NewType(FamilyObj(children),
-                               IsCongruenceLattice),
-                       [children, congs]);
-  return lattice;
+  # Make the object and return
+  po := Digraph(children);
+  SetInNeighbours(po, parents);
+  poset := Objectify(NewType(FamilyObj(children), IsCongruencePoset),
+                     rec(po := po));
+  SetCongruencesOfPoset(poset, congs);
+  SetRange(poset, S);
+  return poset;
 end;
 
+InstallMethod(PosetOfCongruences,
+"for a list or collection",
+[IsListOrCollection],
+function(coll)
+  local congs, nrcongs, children, parents, i, ignore, j, po, poset;
+  congs := AsList(coll);
+  nrcongs := Length(congs);
+
+  # Setup children and parents lists
+  children := [];
+  parents := [];
+  for i in [1 .. nrcongs] do
+    children[i] := Set([]);
+    parents[i] := Set([]);
+  od;
+
+  # Find children of each cong in turn
+  for i in [1 .. nrcongs] do
+    # Ignore self and known parents
+    ignore := BlistList([1 .. nrcongs], [parents[i]]);
+    ignore[i] := true;
+    for j in [1 .. nrcongs] do
+      if not ignore[j] and IsSubrelation(congs[i], congs[j]) then
+        AddSet(children[i], j);
+        AddSet(parents[j], i);
+      fi;
+    od;
+  od;
+
+  # We are done: make the object and return
+  po := Digraph(children);
+  SetInNeighbours(po, parents);
+  poset := Objectify(NewType(FamilyObj(children), IsCongruencePoset),
+                     rec(po := po));
+  SetCongruencesOfPoset(poset, congs);
+  if nrcongs > 0 then
+    SetRange(poset, Range(congs[1]));
+  fi;
+  return poset;
+end);
+
 InstallMethod(LatticeOfLeftCongruences,
 "for a semigroup", [IsSemigroup],
-S -> SEMIGROUPS.LatticeOfXCongruences(S, "Left", rec()));
+S -> LatticeOfLeftCongruences(S, S));
 
 InstallMethod(LatticeOfRightCongruences,
 "for a semigroup", [IsSemigroup],
-S -> SEMIGROUPS.LatticeOfXCongruences(S, "Right", rec()));
+S -> LatticeOfRightCongruences(S, S));
 
 InstallMethod(LatticeOfCongruences,
 "for a semigroup", [IsSemigroup],
-S -> SEMIGROUPS.LatticeOfXCongruences(S, "", rec()));
+S -> LatticeOfCongruences(S, S));
 
 InstallMethod(LatticeOfLeftCongruences,
 "for a semigroup and a multiplicative element collection",
 [IsSemigroup, IsMultiplicativeElementCollection],
 function(S, restriction)
-  if not ForAll(restriction, x -> x in S) then
-    ErrorNoReturn("Semigroups: LatticeOfLeftCongruences: usage,\n",
-                  "<restriction> must only contain elements ",
-                  "in the semigroup <S>,");
-  fi;
-  return SEMIGROUPS.LatticeOfXCongruences(S, "Left",
-                                          rec(restriction := restriction));
+  local poset;
+  poset := SEMIGROUPS.PrincipalXCongruencePoset(S, restriction,
+                                                LeftSemigroupCongruence);
```

Comment:
Why are you using `SEMIGROUPS.PrincipalXCongruencePoset` instead of `PrincipalLeftCongruences(S, restriction)` here?